src/Ormolu/Printer/Meat/Module.hs

```diff
@@ -46,38 +45,11 @@ p_hsModule loc@(L moduleSpan hsModule) = do
     forM_ (sortImports hsmodImports) (located' p_hsmodImport)
     when (not (null hsmodImports) && not (null hsmodDecls)) newline
 
-    forM_ (zip hsmodDecls ((Just <$> drop 1 hsmodDecls) ++ [Nothing])) $
-      \(d, md) -> do
-        case md of
-          Nothing -> located d p_hsDecl
-          Just d' ->
-            if separatedDecls (unL d) (unL d')
-              then line (located d p_hsDecl)
-              else located d p_hsDecl
+    p_groupingHelper
+      (located' p_hsDecl)
+      (\d d' -> separatedDecls (unL d) (unL d'))
```

Comment:
It seems like you only ever call this function with `separatedDecls`, so maybe you should inline it to `p_groupingHelper` for now. Once there's an actual use case for generalising the grouping predicate then you can resurrect that logic.